### PR TITLE
feat(cfn-include): handle resources not in the CloudFormation schema

### LINKED ISSUE #9197: [cfn-include] Support custom CloudFormation resources
Currently, custom resources are not supported; this module errors out if any are used. 

### Use Case
CloudFormation supports custom resources, so this module should as well.

### Proposed Solution

I will be opening a PR to address this.

* [x] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by comcalvi**
Closes  #9197

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/cloudformation-include/test/valid-templates.test.ts [hunk 1]
```diff
@@ -444,10 +444,64 @@ describe('CDK Include', () => {
...
+      "UpdateReplacePolicy": "Retain",
+    }, ResourcePart.CompleteDefinition);
+  });
+
+  test('can ingest a custom resource that uses conditions', () => {
+    const cfnTempalte = includeTestTemplate(stack, 'custom-resource-with-condition.json');
+    const alwaysFalseCondition = cfnTempalte.getCondition('AlwaysFalseCond');
+
+    expect(cfnTempalte.getResource('CustomResource').cfnOptions.condition).toBe(alwaysFalseCondition);
+  });
```
Comment (skinny85) [inline packages/@aws-cdk/cloudformation-include/test/valid-templates.test.ts:499]: I would add a `Condition` attribute to the test above, and get rid of this one.

File: packages/@aws-cdk/cloudformation-include/test/valid-templates.test.ts [hunk 2]
```diff
@@ -444,10 +444,64 @@ describe('CDK Include', () => {
...
+
+  test('throws an exception when an undefined condition is used', () => {
```
Comment (skinny85) [inline packages/@aws-cdk/cloudformation-include/test/valid-templates.test.ts:501]:
```suggestion
  test("throws an exception when a custom resource uses a Condition attribute that doesn't exist in the template", () => {
```

I feel like I've made this comment many times before.

`undefined condition` means something completely different in this context (I would expect a test with that name  to test something like `"Condtion": null` in the template).

File: tools/cfn2ts/lib/codegen.ts [hunk 3]
```diff
@@ -252,37 +252,8 @@ export default class CodeGenerator {
...
-
-    // handle Condition
-    this.code.line('// handle Condition');
-    this.code.openBlock('if (resourceAttributes.Condition)');
-    this.code.line('const condition = options.finder.findCondition(resourceAttributes.Condition);');
-    this.code.openBlock('if (!condition)');
-    this.code.line("throw new Error(`Resource '${id}' uses Condition '${resourceAttributes.Condition}' that doesn't exist`);");
-    this.code.closeBlock();
-    this.code.line('cfnOptions.condition = condition;');
-    this.code.closeBlock();
```
Comment (skinny85) [inline tools/cfn2ts/lib/codegen.ts:255]: No empty line here.

File: tools/cfn2ts/lib/codegen.ts [hunk 4]
```diff
@@ -252,37 +252,8 @@ export default class CodeGenerator {
...
 
+    this.code.line('cfn_parse.handleAttributes(ret, resourceAttributes, id, options.finder);');
```
Comment (skinny85) [inline tools/cfn2ts/lib/codegen.ts:256]: Add an empty line after this line.

File: tools/cfn2ts/lib/codegen.ts [hunk 5]
```diff
@@ -236,10 +236,10 @@ export default class CodeGenerator {
     this.code.openBlock(`public static fromCloudFormation(scope: ${CONSTRUCT_CLASS}, id: string, resourceAttributes: any, options: ${CORE}.FromCloudFormationOptions): ` +
       `${resourceName.className}`);
     this.code.line('resourceAttributes = resourceAttributes || {};');
-    this.code.indent('const cfnParser = new cfn_parse.CfnParser({');
-    this.code.line('finder: options.finder,');
-    this.code.unindent('});');
     if (propsType) {
+      this.code.indent('const cfnParser = new cfn_parse.CfnParser({');
+      this.code.line('finder: options.finder,');
+      this.code.unindent('});');
```
Comment (skinny85) [inline tools/cfn2ts/lib/codegen.ts:242]: How about making `handleAttributes` a method on `CfnParser`? This way, you wouldn't have to change this, and also you wouldn't have to pass `finder` to it (as it already has one).

Thoughts on this idea?

File: packages/@aws-cdk/core/lib/cfn-parse.ts [hunk 6]
```diff
@@ -432,3 +433,39 @@ function specialCaseRefs(value: any): any {
...
+  const dependencies: string[] = Array.isArray(resourceAttributes.DependsOn) ?
+    resourceAttributes.DependsOn : [resourceAttributes.DependsOn];
+  for (const dep of dependencies) {
+    const depResource = finder.findResource(dep);
+    if (!depResource) {
+      throw new Error(`Resource '${logicalId}' depends on '${dep}' that doesn't exist`);
+    }
+    resource.node.addDependency(depResource);
+  }
+  // handle Condition
```
Comment (skinny85) [inline packages/@aws-cdk/core/lib/cfn-parse.ts:463]: Can you actually move handling Condition above DependsOn? Yeah, I know it was the other way originally, but I think Condition first makes more sense. Thanks!

File: packages/@aws-cdk/cloudformation-include/test/valid-templates.test.ts [hunk 7]
```diff
@@ -444,10 +444,64 @@ describe('CDK Include', () => {
...
+
+    expect(stack).toHaveResourceLike('AWS::MyService::AnotherCustom', {
+      "UpdatePolicy": {
+        "AutoScalingReplacingUpdate": {
+          "WillReplace": "false",
+        },
+      },
+      "UpdateReplacePolicy": "Retain",
+    }, ResourcePart.CompleteDefinition);
+  });
```
Comment (skinny85) [inline packages/@aws-cdk/cloudformation-include/test/valid-templates.test.ts:492]: Actually, shouldn't this just check that the resulting template is the same as the input one?

File: packages/@aws-cdk/cloudformation-include/test/nested-stacks.test.ts [hunk 8]
```diff
@@ -387,8 +387,6 @@ describe('CDK Include', () => {
       ],
       "UpdateReplacePolicy": "Retain",
     }, ResourcePart.CompleteDefinition);
-
-    cfnTemplate.getNestedStack('AnotherChildStack');
   });
```
[the remark this answers is not in the document]
Comment (comcalvi) [reply packages/@aws-cdk/cloudformation-include/test/nested-stacks.test.ts:390]: there is no assertion here; those removed lines don't do anything.

[the remark this answers is not in the document]
Comment (skinny85) [reply packages/@aws-cdk/cloudformation-include/test/nested-stacks.test.ts:390]: Actually, they did. They tested that a forward-declared reference on Stack resource in the template is resolved correctly (that is, `cfnTemplate.getNestedStack('AnotherChildStack')` does not throw). In fact, this test failed on an iteration of the nested stack production code, and that's why it was added here.

So, I wouldn't touch them in this PR.

File: packages/@aws-cdk/cloudformation-include/test/valid-templates.test.ts (hunk 2 above)
Comment (comcalvi) [reply packages/@aws-cdk/cloudformation-include/test/valid-templates.test.ts:501]: Oh, I see...my bad, I've been misunderstanding the meaning of `undefined` in this context.

File: packages/@aws-cdk/cloudformation-include/lib/cfn-include.ts [hunk 9]
```diff
@@ -356,10 +348,32 @@ export class CfnInclude extends core.CfnElement {
...
+
+      l1Instance = this.nestedStacksToInclude[logicalId]
+        ? this.createNestedStack(logicalId, finder)
+        : jsClassFromModule.fromCloudFormation(this, logicalId, resourceAttributes, options);
+    }
 
-    const l1Instance = this.nestedStacksToInclude[logicalId]
-      ? this.createNestedStack(logicalId, finder)
-      : jsClassFromModule.fromCloudFormation(this, logicalId, resourceAttributes, options);
+    else {
```
Comment (skinny85) [inline packages/@aws-cdk/cloudformation-include/lib/cfn-include.ts:364]: What's with the weird formatting? it should be `} else {`.

File: packages/@aws-cdk/cloudformation-include/lib/cfn-include.ts [hunk 10]
```diff
@@ -356,10 +348,32 @@ export class CfnInclude extends core.CfnElement {
...
     };
+    const l1ClassFqn = cfn_type_to_l1_mapping.lookup(resourceAttributes.Type);
+    let l1Instance;
+
+    if (l1ClassFqn) {
+      const [moduleName, ...className] = l1ClassFqn.split('.');
+      const module = require(moduleName);  // eslint-disable-line @typescript-eslint/no-require-imports
+      const jsClassFromModule = module[className.join('.')];
+
+      l1Instance = this.nestedStacksToInclude[logicalId]
```
Comment (skinny85) [inline packages/@aws-cdk/cloudformation-include/lib/cfn-include.ts:359]: I found a bug!

The following test:

```ts
  test('throws an exception when and ID was passed in nestedStacks that is a resource type not in the CloudFormation schema', () => {
    expect(() => {
      new inc.CfnInclude(stack, 'Template', {
        templateFile: testTemplateFilePath('custom-resource.json'),
        nestedStacks: {
          'CustomResource': {
            templateFile: testTemplateFilePath('whatever.json'),
          },
        },
      });
    }).toThrow(/Nested Stack with logical ID 'CustomResource' is not an AWS::CloudFormation::Stack resource/);
  });
```

where `custom-resource.json` is:

```json
{
  "Resources": {
    "CustomResource": {
      "Type": "AWS::CustomResource::Type"
    }
  }
}
```

fails with: `Received function did not throw`.

I'll leave it to you to figure out what happened, and fix it 🙂.

File: packages/@aws-cdk/cloudformation-include/README.md [hunk 11]
```diff
@@ -106,6 +106,9 @@ const bucket = s3.Bucket.fromBucketName(this, 'L2Bucket', cfnBucket.ref);
 // bucket is of type s3.IBucket
 ```
 
+Note that [Custom Resources](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-cfn-customresource.html) will be of type
+CfnResource, and hence won't need to be casted. This holds for any resource that isn't in the CloudFormation schema.
```
Comment (skinny85) [inline packages/@aws-cdk/cloudformation-include/README.md:110]: Ends sentences on line endings (Markdown should be readable in plaintext format as well):

```suggestion
Note that [Custom Resources](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-cfn-customresource.html)
will be of type CfnResource,
and hence won't need to be casted.
This holds for any resource that isn't in the CloudFormation schema.
```

File: packages/@aws-cdk/core/lib/cfn-parse.ts [hunk 12]
```diff
@@ -110,7 +111,7 @@ export class FromCloudFormation {
 
   public static getCfnTag(tag: any): CfnTag {
     return tag == null
-      ? { } as any // break the type system - this should be detected at runtime by a tag validator
+      ? {} as any // break the type system - this should be detected at runtime by a tag validator
```
Comment (skinny85) [inline packages/@aws-cdk/core/lib/cfn-parse.ts:114]: No reason to change this :)

File: packages/@aws-cdk/core/lib/cfn-parse.ts [hunk 13]
```diff
@@ -294,6 +295,43 @@ export class CfnParser {
     return cfnValue;
   }
 
+  public handleAttributes(resource: CfnResource, resourceAttributes: any, logicalId: string): void {
+
```
Comment (skinny85) [inline packages/@aws-cdk/core/lib/cfn-parse.ts:299]: Unnecessary empty line.

File: packages/@aws-cdk/core/lib/cfn-parse.ts [hunk 14]
```diff
@@ -294,6 +295,43 @@ export class CfnParser {
...
+    const finder = this.options.finder;
+    const cfnParser = new CfnParser({
```
Comment (skinny85) [inline packages/@aws-cdk/core/lib/cfn-parse.ts:301]: Uhm. No reason to create a `cnfParser`, we are inside one :). So just call `this. parseCreationPolicy`, etc.

File: packages/@aws-cdk/core/lib/cfn-parse.ts [hunk 15]
```diff
@@ -432,3 +470,5 @@ function specialCaseRefs(value: any): any {
 function undefinedIfAllValuesAreEmpty(object: object): object | undefined {
   return Object.values(object).some(v => v !== undefined) ? object : undefined;
 }
+
```
Comment (skinny85) [inline packages/@aws-cdk/core/lib/cfn-parse.ts:473]: Nope :)

File: packages/@aws-cdk/core/lib/cfn-parse.ts [hunk 16]
```diff
@@ -408,8 +446,8 @@ export class CfnParser {
 
     const key = objectKeys[0];
     return key === 'Ref' || key.startsWith('Fn::') ||
-        // special intrinsic only available in the 'Conditions' section
-        (this.options.context === CfnParsingContext.CONDITIONS && key === 'Condition')
+      // special intrinsic only available in the 'Conditions' section
+      (this.options.context === CfnParsingContext.CONDITIONS && key === 'Condition')
```
Comment (skinny85) [inline packages/@aws-cdk/core/lib/cfn-parse.ts:450]: No reason to change this.

File: packages/@aws-cdk/core/lib/cfn-parse.ts [hunk 17]
```diff
@@ -294,6 +295,43 @@ export class CfnParser {
...
+    const cfnParser = new CfnParser({
+      finder,
+    });
+
+    const cfnOptions = resource.cfnOptions;
+
+    cfnOptions.creationPolicy = cfnParser.parseCreationPolicy(resourceAttributes.CreationPolicy);
+    cfnOptions.updatePolicy = cfnParser.parseUpdatePolicy(resourceAttributes.UpdatePolicy);
+    cfnOptions.deletionPolicy = cfnParser.parseDeletionPolicy(resourceAttributes.DeletionPolicy);
+    cfnOptions.updateReplacePolicy = cfnParser.parseDeletionPolicy(resourceAttributes.UpdateReplacePolicy);
```
Comment (skinny85) [inline packages/@aws-cdk/core/lib/cfn-parse.ts:310]: I believe `parseCreationPolicy`, `parseUpdatePolicy` and `parseDeletionPolicy` can now be made private in `CfnParser`.

File: packages/@aws-cdk/cloudformation-include/test/valid-templates.test.ts [hunk 18]
```diff
@@ -444,10 +444,28 @@ describe('CDK Include', () => {
...
+    const cfnTemplate = includeTestTemplate(stack, 'custom-resource-with-attributes.json');
+    expect(stack).toMatchTemplate(loadTestFileToJsObject('custom-resource-with-attributes.json'));
+
+    const alwaysFalseCondition = cfnTemplate.getCondition('AlwaysFalseCond');
+    expect(cfnTemplate.getResource('CustomBucket').cfnOptions.condition).toBe(alwaysFalseCondition);
+  });
+
+  test("throws an exception when a custom resource uses a Condition attribute that doesn't exist in the template", () => {
     expect(() => {
-      includeTestTemplate(stack, 'non-existent-resource-type.json');
```
Comment (skinny85) [inline packages/@aws-cdk/cloudformation-include/test/valid-templates.test.ts:449]: I think that means you can remove `non-existent-resource-type.json` completely...?

File: packages/@aws-cdk/cloudformation-include/test/valid-templates.test.ts [hunk 19]
```diff
@@ -444,10 +444,28 @@ describe('CDK Include', () => {
...
+      "CustomProp": "CustomValue",
+      "CustomFuncProp": {
+        "Ref": "AWS::NoValue",
+      },
+    });
+  });
+
+  test('can include a template with a custom resource that uses attributes', () => {
+    const cfnTemplate = includeTestTemplate(stack, 'custom-resource-with-attributes.json');
+    expect(stack).toMatchTemplate(loadTestFileToJsObject('custom-resource-with-attributes.json'));
```
Comment (skinny85) [inline packages/@aws-cdk/cloudformation-include/test/valid-templates.test.ts:459]:
```suggestion
    expect(stack).toMatchTemplate(
      loadTestFileToJsObject('custom-resource-with-attributes.json'),
    );
```

File: packages/@aws-cdk/cloudformation-include/test/valid-templates.test.ts [hunk 20]
```diff
@@ -444,10 +444,28 @@ describe('CDK Include', () => {
     });
   });
 
-  test("throws an exception when encountering a Resource type it doesn't recognize", () => {
+  test('can include a template with a custom resource that has custom properties', () => {
+    includeTestTemplate(stack, 'custom-resource-properties.json');
+    expect(stack).toHaveResourceLike('AWS::MyService::Custom', {
```
Comment (skinny85) [inline packages/@aws-cdk/cloudformation-include/test/valid-templates.test.ts:449]: Can we just add a property to `CustomResource` in `custom-resource-with-attributes.json`, and get rid of this test completely?

File: packages/@aws-cdk/core/lib/cfn-parse.ts (hunk 15 above)
Comment (comcalvi) [reply packages/@aws-cdk/core/lib/cfn-parse.ts:473]: I have no idea how that line got there...I really don't really remember adding that, and I'm surprised the linter didn't catch the triple blank line.

File: packages/@aws-cdk/core/lib/cfn-parse.ts (hunk 17 above)
Comment (comcalvi) [reply packages/@aws-cdk/core/lib/cfn-parse.ts:310]: `parseUpdateReplacePolicy` as well?

**Pull request merged**